src/main/scala/org/broadinstitute/dsde/workbench/sam/util/DistributedLock.scala

```diff
@@ -0,0 +1,105 @@
+  implicit val cs = IO.contextShift(ExecutionContext.global)
+  implicit val timer = IO.timer(ExecutionContext.global)
+
+  val executor = new Executor {
+    override def execute(command: Runnable): Unit = ExecutionContext.global.execute(command)
+  }
+
+  def withLock(lockPath: LockPath)(implicit googleFirestoreOps: GoogleFirestoreOps[IO]): Resource[IO, Unit] = Resource.make{
+    retry(getLock(lockPath), 5 seconds, 5)
+  }(_ => releaseLock(lockPath))
+
+  private[util] def getLock[A](lockPath: LockPath)(implicit googleFirestoreOps: GoogleFirestoreOps[IO]): IO[Unit] =
+    googleFirestoreOps.transaction { (db, tx) =>
+      val docRef = db.collection(lockPath.collectionName.asString).document(lockPath.document.asString)
+      val res = for {
+        lockStatus <- getLockStatus(tx, docRef)
+        _ <- lockStatus match {
+          case Available => setLock(tx, docRef, lockPath.expiresIn)
+          case Locked => IO.raiseError(new WorkbenchException("can't get lock"))
+        }
+      } yield ()
+      res.unsafeRunSync()
+    }
+
+  private[util] def getLockStatus(tx: Transaction, documentReference: DocumentReference): IO[LockStatus] = {
+    IO.async[DocumentSnapshot] { cb =>
+       Try(tx.get(documentReference).get()).fold(t => cb(Left(t)), doc => cb(Right(doc)))
+      }
+      .flatMap{x =>
+        Option(x.getBoolean(ISLOCKED)).map(_.asInstanceOf[Boolean]) match{
+          case Some(true) =>
+            val expiresIn = Option(x.getLong(EXPIRESIN))
+            expiresIn match{
+              case Some(e) =>
+                timer.clock.realTime(EXPIRESINTIMEUNIT).map{
+                  currentTime =>
+                    if(e < currentTime)
+                      Available
+                    else
+                      Locked
+                }
+              case None => IO.raiseError(new WorkbenchException("Lock expiration time is missing"))
+            }
+          case None | Some(false) => IO.pure(Available)
+        }
+      }
+  }
+
+  private[util] def releaseLock(lockPath: LockPath)(implicit googleFirestoreOps: GoogleFirestoreOps[IO]): IO[Unit] = {
+    googleFirestoreOps.transaction{
+      (db, tx) =>
+        val docRef = db.collection(lockPath.collectionName.asString).document(lockPath.document.asString)
+        tx.set(docRef, Map(ISLOCKED -> false).asJava)
+    }
+  }
+
+  private[util] def setLock(tx: Transaction, documentReference: DocumentReference, expiresIn: ExpiresIn): IO[Unit] =
+    for {
+      currentTime <- timer.clock.realTime(EXPIRESINTIMEUNIT)
+      data = Map(
+        EXPIRESIN -> (currentTime + expiresIn.asDuration.toMillis),
+        ISLOCKED -> true
+      )
+      _ <- IO(tx.set(documentReference, data.asJava))
+    } yield ()
+
+  def retry[A](ioa: IO[A], initialDelay: FiniteDuration, maxRetries: Int)
+                         (implicit timer: Timer[IO]): IO[A] = {
+    ioa.handleErrorWith {
+      case error: WorkbenchException =>
+        IO.raiseError(error)
```

Comment:
I am not quite following how the retry will work. retry is called with getLock which throws WorkbenchException if it can't get the lock... wouldn't that be handled by this case and thus not retry?